tests/stages/test_discovery.py

```diff
@@ -66,69 +126,95 @@ def test_should_find_invalidated_test_dependencies(self):
-                    "some_other_unrelated_file.txt": "A",
-                },
-            ),
-            steps=self.steps,
-        )
-        assert 1 == len(project_executions)
-        assert project_executions.pop().project == load_project(
-            root_test_path.parent,
-            Path("tests/projects/job/deployment/project.yml"),
-            strict=False,
+    def test_stage_with_files_changed(self):
+        project_executions = self._helper_find_projects_to_execute(
+            files_touched={
+                "tests/projects/job/deployment/project.yml": "M",
+            },
         )
+        assert len(project_executions) == 1
+        job_execution = next(p for p in project_executions if p.project.name == "job")
+        assert not job_execution.cached
 
-    def test_build_project_executions_when_all_files_filtered(self):
-        project_paths = [
-            "tests/projects/job/deployment/project.yml",
-            "tests/projects/service/deployment/project.yml",
-            "tests/projects/sbt-service/deployment/project.yml",
-        ]
-        projects = set(load_projects(root_test_path.parent, project_paths))
-        project_executions = find_projects_to_execute(
-            logger=self.logger,
-            all_projects=projects,
-            stage=TestStage.build().name,
-            changeset=Changeset(
-                sha="a git SHA",
-                _files_touched={
+    def test_stage_with_files_changed_and_existing_cache(self):
+        with _caching_for(project="job"):
+            project_executions = self._helper_find_projects_to_execute(
+                files_touched={
+                    "tests/projects/job/deployment/project.yml": "M",
+                },
+            )
+            assert len(project_executions) == 1
+            job_execution = next(
+                p for p in project_executions if p.project.name == "job"
+            )
+            assert job_execution.cached
+
+    def test_stage_with_files_changed_but_filtered(self):
+        with _caching_for(project="job"):
+            project_executions = self._helper_find_projects_to_execute(
+                files_touched={
                     "tests/projects/job/deployment/project.yml": "D",
                 },
-            ),
-            steps=self.steps,
-        )
-        assert 1 == len(project_executions)
-        execution = project_executions.pop()
-        assert execution.cache_key == "a git SHA"
-        assert execution.project == load_project(
-            root_test_path.parent,
-            Path("tests/projects/job/deployment/project.yml"),
-            strict=False,
+            )
+            assert len(project_executions) == 1
+            job_execution = next(
+                p for p in project_executions if p.project.name == "job"
+            )
+            assert not job_execution.cached
+
+    def test_stage_with_build_dependency_changed(self):
```

Comment:
Do we execute all stages for the build dependency now?

Reply:
No, only stages that match the dependency stage are executed:
* when running the build stage, trigger projects that are build dependencies
* when running the test stage, trigger projects that are test dependencies
* when running the deploy stage, trigger projects that are deploy dependencies
* and so on

The `test_` prefix might be confusing you here, it doesn't refer to the "test" stage but to the fact that this function is a unit test (and all unit tests start with `test_" prefix) 🙂

Reply:
Yeah that's what I thought.. So currently we have 2 bugs in mpyl. One is that the dependencies are not taken into account when caching (which this PR fixes), and the other is that build dependencies only trigger the build stage, not all downstream stages. The second bug was introduced in the configurable stages PR, but it is a serious bug in my opinion, because currently a lot of projects depend on this in the monorepo (that's why it was introduced to mpyl as well).

My idea would be to extend the stages definition to account for this type of dependencies. Normal behaviour would be what we have now, but for build, I would add in the run_properties and in the stages schema something like:

```yaml
stages:
  - name: 'build'
    icon: '🏗️'
    dependency_strategy: 'BuildAll'
  - name: 'test'
    icon: '📋'
    # dependency_strategy: 'BuildStage' (default)
  - name: 'deploy'
    icon: '🚀'
  - name: 'postdeploy'
    icon: '🦺️'
```

Reply:
Fully agree with you! The only thing I'd challenge is the categorizing: I don't think this behaviour is per se a bug, it seemed to have been changed deliberately by a previous release. I do agree it breaks some expectations on our side though 👍 

I think your suggestion makes a lot of sense, maybe we can book a session to go over these rules in more detail?